ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
+            // For example {S} will be reshaped to {S, 1}.
+            arg1_shape_tmp.insert(arg1_shape_tmp.end(), 1);
+            arg1_rank = arg1_shape_tmp.size();
         }
-        else if (arg1_rank == 1)
+
+        // Check matrices dimensions compatibility,
+        // COL_INDEX_DIM of the first matrix has to match ROW_INDEX_DIM of the second matrix.
+        auto merged_dimension = Dimension::dynamic();
+        auto arg0_col_dim = arg0_shape_tmp[arg0_rank - 1];
+        auto arg1_row_dim = arg1_shape_tmp[arg1_rank - 2];
+        NGRAPH_CHECK(Dimension::merge(merged_dimension, arg0_col_dim, arg1_row_dim) ||
+                         arg0_col_dim.is_dynamic() || arg1_row_dim.is_dynamic(),
+                     "Incompatible MatMul matrix dimension. ",
+                     "First input dimension=",
+                     arg0_col_dim,
+                     " at COL_INDEX_DIM=",
+                     (arg0_rank - 1),
+                     " doesn't match the second input dimension=",
+                     arg1_row_dim,
+                     " at ROW_INDEX_DIM=",
+                     (arg1_rank - 2));
+
+        // 3. If ranks of input arguments are different after steps 1 and 2,
+        // the smaller tensor is unsqueezed from the left side of the shape
+        // by necessary number of axes to make both shapes of the same rank.
+        std::vector<Dimension> small_batch_matrix =
+            arg0_rank > arg1_rank ? arg1_shape_tmp : arg0_shape_tmp;
+        std::vector<Dimension> big_batch_matrix =
+            arg0_rank > arg1_rank ? arg0_shape_tmp : arg1_shape_tmp;
+        std::vector<Dimension> output_shape(big_batch_matrix);
+
+        if (arg0_rank != arg1_rank)
         {
-            // i.e., arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
-            NGRAPH_CHECK(arg1_shape_update[0] == arg0_shape_update[arg0_rank - 1],
-                         "Incompatible arg shapes");
-            arg0_shape_update.erase(arg0_shape_update.begin() + arg0_rank - 1);
-            output_shape = arg0_shape_update;
+            // Expand small_batch_matrix (with 1) to have the same rank as big_batch_matrix
+            size_t delta_rank = big_batch_matrix.size() - small_batch_matrix.size();
+            small_batch_matrix.insert(small_batch_matrix.begin(), delta_rank, 1);
         }
-        else if (arg0_rank == 2 && arg1_rank == 2)
+
+        // 4. Usual rules of the broadcasting are applied for batch dimensions.
+        // Broadcast all batches (last two dimensions represent matrix),
+        // expand dim with value 1 to bigger dim if dimensions are not equal.
+        for (auto i = 0; i < big_batch_matrix.size() - 2; i++)
         {
-            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
-            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
+            // Dimension with value 1 can be expanded to any bigger
```

Comment:
This comment is not so relevant to the next code without additional explanations. It is more suitable for case `if (lower_bound <= 1)`.

Reply:
Done